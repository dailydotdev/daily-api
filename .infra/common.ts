export interface WorkerArgs {
  enableMessageOrdering?: boolean;
  ackDeadlineSeconds?: number;
  expirationPolicy?: {
    ttl: string;
  };
  deadLetterPolicy?: {
    deadLetterTopic: string;
    maxDeliveryAttempts: number;
  };
}

export interface Worker {
  topic: string;
  subscription: string;
  args?: WorkerArgs;
}

export const digestDeadLetter = 'api.v1.personalized-digest-email-dead-letter';

export const workers: Worker[] = [
  {
    topic: 'views',
    subscription: 'add-views-v2',
  },
  {
    topic: 'user-updated',
    subscription: 'api.user-updated-cio',
  },
  {
    topic: 'user-deleted',
    subscription: 'api.user-deleted-cio',
  },
  {
    topic: 'comment-upvoted',
    subscription: 'comment-upvoted-rep',
  },
  {
    topic: 'comment-upvote-canceled',
    subscription: 'comment-upvote-canceled-rep',
  },
  {
    topic: 'api.v1.post-visible',
    subscription: 'api.post-scout-matched-slack-v2',
  },
  {
    topic: 'api.v1.post-visible',
    subscription: 'api.post-freeform-images',
  },
  {
    topic: 'api.v1.post-content-edited',
    subscription: 'api.post-edited-freeform-images',
  },
  {
    topic: 'comment-commented',
    subscription: 'comment-commented-slack-message',
  },
  {
    topic: 'post-upvoted',
    subscription: 'post-upvoted-rep',
  },
  {
    topic: 'post-upvote-canceled',
    subscription: 'post-upvote-canceled-rep',
  },
  {
    topic: 'api.v1.post-downvoted',
    subscription: 'api.post-downvoted-rep',
  },
  {
    topic: 'api.v1.post-downvote-canceled',
    subscription: 'api.post-downvote-canceled-rep',
  },
  {
    topic: 'post-commented',
    subscription: 'api.post-commented-images',
  },
  {
    topic: 'post-commented',
    subscription: 'post-commented-redis',
  },
  {
    topic: 'post-commented',
    subscription: 'post-commented-slack-message',
  },
  {
    topic: 'post-commented',
    subscription: 'api.vordr-post-comment-prevented',
  },
  {
    topic: 'post-upvoted',
    subscription: 'post-upvoted-redis',
  },
  {
    topic: 'post-banned-or-removed',
    subscription: 'post-banned-rep',
  },
  {
    topic: 'post-banned-or-removed',
    subscription: 'api.post-deleted-comments-cleanup',
  },
  {
    topic: 'pub-request',
    subscription: 'pub-request-rep',
  },
  {
    topic: 'username-changed',
    subscription: 'username-changed-api',
  },
  {
    topic: 'update-comments',
    subscription: 'update-comments-mention',
  },
  {
    topic: 'api.changes',
    subscription: 'api-cdc',
    args: { enableMessageOrdering: true },
  },
  {
    topic: 'api.v1.cdc-notifications',
    subscription: 'api.cdc-notifications',
  },
  {
    topic: 'api.v1.new-notification',
    subscription: 'api.new-notification-real-time',
  },
  {
    topic: 'api.v1.new-notification',
    subscription: 'api.new-notification-mail',
  },
  {
    topic: 'api.v1.new-notification',
    subscription: 'api.new-notification-push',
  },
  {
    topic: 'api.v1.source-privacy-updated',
    subscription: 'api.source-privacy-updated',
  },
  {
    topic: 'api.v1.content-image-deleted',
    subscription: 'api.delete-cloudinary-image',
  },
  {
    topic: 'api.v1.comment-edited',
    subscription: 'api.comment-edited-images',
  },
  {
    topic: 'api.v1.comment-deleted',
    subscription: 'api.comment-deleted-images',
  },
  // Notifications
  {
    topic: 'community-link-rejected',
    subscription: 'api.community-picks-failed-notification',
  },
  {
    topic: 'community-link-access',
    subscription: 'api.community-picks-granted-notification',
  },
  {
    topic: 'post-commented',
    subscription: 'api.article-new-comment-notification.post-commented',
  },
  {
    topic: 'comment-commented',
    subscription: 'api.article-new-comment-notification.comment-commented',
  },
  {
    topic: 'api.v1.post-bookmark-reminder',
    subscription: 'api.post-bookmark-reminder-notification',
  },
  {
    topic: 'comment-commented',
    subscription: 'api.comment-commented-images',
  },
  {
    topic: 'post-upvoted',
    subscription: 'api.article-upvote-milestone-notification',
  },
  {
    topic: 'send-analytics-report',
    subscription: 'api.article-analytics-notification',
  },
  {
    topic: 'post-banned-or-removed',
    subscription: 'api.article-report-approved-notification',
  },
  {
    topic: 'pub-request',
    subscription: 'api.source-request-notification',
  },
  {
    topic: 'api.v1.new-post-mention',
    subscription: 'api.post-mention-notification',
  },
  {
    topic: 'api.v1.new-comment-mention',
    subscription: 'api.comment-mention-notification',
  },
  {
    topic: 'user-reputation-updated',
    subscription: 'api.user-reputation-updated-notification',
  },
  {
    topic: 'comment-commented',
    subscription: 'api.comment-reply-notification',
  },
  {
    topic: 'comment-upvoted',
    subscription: 'api.comment-upvote-milestone-notification',
  },
  {
    topic: 'api.v1.post-visible',
    subscription: 'api.post-added-notification-v2',
  },
  {
    topic: 'api.v1.member-joined-source',
    subscription: 'api.member-joined-source-notification',
  },
  {
    topic: 'api.v1.user-created',
    subscription: 'api.user-created-cio',
  },
  {
    topic: 'api.v1.banner-added',
    subscription: 'api.banner-added',
  },
  {
    topic: 'api.v1.banner-deleted',
    subscription: 'api.banner-deleted',
  },
  {
    topic: 'yggdrasil.v1.content-published',
    subscription: 'api.content-published',
  },
  {
    topic: 'api.v1.source-member-role-changed',
    subscription: 'api.source-member-role-changed-notification',
  },
  {
    topic: 'analytics-api.v1.experiment-allocated',
    subscription: 'api.experiment-allocated',
  },
  {
    topic: 'api.v1.source-created',
    subscription: 'api.source-created-squad-user-action',
  },
  {
    topic: 'api.v1.post-collection-updated',
    subscription: 'api.post-collection-updated-notification',
  },
  {
    topic: 'api.v1.user-readme-updated',
    subscription: 'api.user-readme-images',
  },
  {
    topic: 'api.v1.user-created',
    subscription: 'api.user-created-personalized-digest-send-type',
  },
  {
    topic: 'api.v1.comment-downvoted',
    subscription: 'api.comment-downvoted-rep',
  },
  {
    topic: 'api.v1.comment-downvote-canceled',
    subscription: 'api.comment-downvote-canceled-rep',
  },
  {
    topic: 'api.v1.squad-public-request',
    subscription: 'api.v1.squad-public-request-notification',
  },
  {
    topic: 'api.v1.user-streak-updated',
    subscription: 'api.user-streak-updated-cio',
  },
  {
    topic: 'api.v1.user-streak-updated',
    subscription: 'api.user-streak-reset-notification',
  },
  {
    topic: 'api.v1.squad-featured-updated',
    subscription: 'api.squad-featured-updated-notification',
  },
  {
    topic: 'api.v1.post-visible',
    subscription: 'api.post-added-slack-channel-send',
  },
  {
    topic: 'api.v1.post-visible',
    subscription: 'api.vordr-post-prevented',
  },
  {
    topic: 'api.v1.user-company-approved',
    subscription: 'api.user-company-approved-cio',
  },
  {
    topic: 'api.v1.post-visible',
    subscription: 'api.post-added-user-notification',
  },
  {
    topic: 'api.v1.user-top-reader',
    subscription: 'api.user-top-reader-added',
  }
];

export const personalizedDigestWorkers: Worker[] = [
  {
    topic: 'api.v1.generate-personalized-digest',
    subscription: 'api.personalized-digest-email',
  },
  {
    topic: digestDeadLetter,
    subscription: 'api.personalized-digest-email-dead-letter-log',
  },
];
