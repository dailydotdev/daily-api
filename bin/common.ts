import fastq from 'fastq';
import { ReadStream } from 'fs';

const QUEUE_CONCURRENCY = 1;

export const runInQueueStream = async <T>(
  stream: ReadStream,
  callback: (props: T) => Promise<void>,
  queue = QUEUE_CONCURRENCY,
) => {
  let insertCount = 0;
  const insertQueue = fastq.promise(async (props: T) => {
    await callback(props);

    insertCount += 1;
  }, queue);

  stream.on('data', insertQueue.push);

  await new Promise((resolve, reject) => {
    stream.on('error', reject);
    stream.on('end', resolve);
  });
  await insertQueue.drained();
  console.log('update finished with a total of: ', insertCount);
};
