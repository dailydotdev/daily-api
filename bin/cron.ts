import '../src/config';
import cron from '../src/cron';
import { createOrGetConnection } from '../src/db';

const key = process.argv[2];
const selected = cron.get(key);
if (!selected) {
  console.error(`cron "${key}" does not exist!`);
  process.exit(-1);
}

console.log(`starting ${key}`);
createOrGetConnection()
  .then((con) => selected.handler(con, ...process.argv.slice(3)))
  .then(() => {
    console.log('done');
    process.exit();
  })
  .catch((err) => {
    console.error(err);
    process.exit(-1);
  });
