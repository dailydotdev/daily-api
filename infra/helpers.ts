import * as pulumi from '@pulumi/pulumi';
import * as gcp from '@pulumi/gcp';
import * as inputs from '@pulumi/gcp/types/input';

function camelToUnderscore(key: string): string {
  const result = key.replace(/([A-Z])/g, ' $1');
  return result.split(' ').join('_').toUpperCase();
}

export const config = new pulumi.Config();
export const stack = pulumi.getStack();
export const location = gcp.config.region || 'us-central1';

export const infra = new pulumi.StackReference(`idoshamun/infra/${stack}`);

export function createEnvVarsFromSecret(
  prefix: string,
): pulumi.Input<inputs.cloudrun.ServiceTemplateSpecContainerEnv>[] {
  const envVars = config.requireObject<Record<string, string>>('env');
  return Object.keys(envVars).map(
    (key): pulumi.Input<inputs.cloudrun.ServiceTemplateSpecContainerEnv> => {
      const secret = new gcp.secretmanager.Secret(`${prefix}-secret-${key}`, {
        secretId: `${prefix}-secret-${key}`,
        replication: { automatic: true },
      });

      const version = new gcp.secretmanager.SecretVersion(
        `${prefix}-sv-${key}`,
        {
          enabled: true,
          secret: secret.name,
          secretData: envVars[key],
        },
      );
      return {
        name: camelToUnderscore(key),
        value: pulumi
          .all([secret.secretId, version.id])
          .apply(
            ([name, version]) =>
              `gcp:///${name}/${version.split('/').reverse()[0]}`,
          ),
      };
    },
  );
}

export function serviceAccountToMember(
  serviceAccount: gcp.serviceaccount.Account,
): pulumi.Output<string> {
  return serviceAccount.email.apply((email) => `serviceAccount:${email}`);
}

export type IAMRole = { name: string; role: string };

export function addIAMRolesToServiceAccount(
  prefix: string,
  roles: IAMRole[],
  serviceAccount: gcp.serviceaccount.Account,
): void {
  const member = serviceAccountToMember(serviceAccount);
  roles.forEach(
    (role) =>
      new gcp.projects.IAMMember(`${prefix}-iam-${role.name}`, {
        role: role.role,
        member,
      }),
  );
}
