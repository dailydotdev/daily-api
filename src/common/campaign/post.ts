import { ValidationError } from 'apollo-server-errors';
import { AuthContext } from '../../Context';
import {
  ArticlePost,
  CampaignPost,
  CampaignState,
  CampaignType,
  Post,
  PostType,
  type ConnectionManager,
  type FreeformPost,
  type SharePost,
} from '../../entity';
import { getPostPermalink } from '../../schema/posts';
import {
  type GetCampaignListResponse,
  type GetCampaignResponse,
} from '../../integrations/skadi';
import type { Connection } from 'graphql-relay';
import { In } from 'typeorm';
import { mapCloudinaryUrl } from '../cloudinary';
import { pickImageUrl } from '../post';
import { NotFoundError } from '../../errors';
import { debeziumTimeToDate, systemUser, updateFlagsStatement } from '../utils';
import { getDiscussionLink } from '../links';
import { skadiApiClientV1 } from '../../integrations/skadi/api/v1/clients';
import { largeNumberFormat } from '../devcard';
import { formatMailDate, addNotificationEmailUtm } from '../mailing';
import { truncatePostToTweet } from '../twitter';
import type { TemplateDataFunc } from '../../workers/newNotificationV2Mail';
import { usdToCores } from '../number';

import {
  startCampaignTransferCores,
  stopCampaignTransferCores,
  type StartCampaignMutationArgs,
  type StopCampaignProps,
} from './common';
import { randomUUID } from 'crypto';
import {
  UserTransaction,
  UserTransactionProcessor,
  UserTransactionStatus,
  UserTransactionType,
} from '../../entity/user/UserTransaction';
import { addDays } from 'date-fns';
import { skadiApiClientV2 } from '../../integrations/skadi/api/v2/clients';

export interface GQLPromotedPost
  extends Omit<
    GetCampaignResponse,
    'spend' | 'budget' | 'startedAt' | 'endedAt'
  > {
  spend: number;
  budget: number;
  startedAt: Date;
  endedAt: Date;
}

export const validatePostBoostPermissions = async (
  ctx: AuthContext,
  postId: string,
): Promise<Pick<Post, 'id' | 'flags'>> => {
  const { userId } = ctx;

  return ctx.con.getRepository(Post).findOneOrFail({
    select: ['id', 'flags'],
    where: [
      { id: postId, authorId: userId },
      { id: postId, scoutId: userId },
    ],
  });
};

export const checkPostAlreadyBoosted = (post: Pick<Post, 'flags'>): void => {
  if (!!post.flags?.campaignId) {
    throw new ValidationError('Post is already boosted');
  }
};

interface CampaignBoostedPost
  extends Pick<Post, 'id' | 'shortId' | 'title' | 'slug'> {
  image: string;
  permalink: string;
  engagements: number;
  commentsPermalink?: string;
}

export interface GQLBoostedPost {
  campaign: GQLPromotedPost;
  post: CampaignBoostedPost;
}

interface GetBoostedPost extends CampaignBoostedPost {
  type: PostType;
  sharedTitle?: string;
  sharedImage?: string;
  views: number;
  upvotes: number;
  comments: number;
}

const getBoostedPostBuilder = (con: ConnectionManager, alias = 'p1') =>
  con
    .getRepository(Post)
    .createQueryBuilder(alias)
    .select(`"${alias}".id`, 'id')
    .addSelect(`"${alias}"."shortId"`, 'shortId')
    .addSelect(`"${alias}".slug`, 'slug')
    .addSelect(`"${alias}".image`, 'image')
    .addSelect(`"${alias}".title`, 'title')
    .addSelect(`"${alias}".type`, 'type')
    .addSelect(`"${alias}".upvotes::int`, 'upvotes')
    .addSelect(`"${alias}".comments::int`, 'comments')
    .addSelect(`"${alias}".views::int`, 'views')
    .addSelect('p2.title', 'sharedTitle')
    .addSelect('p2.image', 'sharedImage')
    .leftJoin(Post, 'p2', `"${alias}"."sharedPostId" = p2.id`);

export const getBoostedPost = async (
  con: ConnectionManager,
  id: string,
): Promise<GetBoostedPost> => {
  const result = await getBoostedPostBuilder(con).where({ id }).getRawOne();

  if (!result) {
    throw new NotFoundError('Post does not exist');
  }

  return result;
};

export const getFormattedBoostedPost = (
  post: GetBoostedPost,
): GQLBoostedPost['post'] => {
  const { id, shortId, sharedImage, sharedTitle, slug } = post;
  let image: string | undefined = post.image;
  let title = post.title;

  if (post.type === PostType.Share) {
    image = sharedImage;
    title = title || sharedTitle;
  }

  return {
    id,
    slug,
    shortId,
    title,
    image: mapCloudinaryUrl(image) ?? pickImageUrl({ createdAt: new Date() }),
    permalink: getPostPermalink({ shortId }),
    commentsPermalink: post.slug ? getDiscussionLink(post.slug) : undefined,
    engagements: post.comments + post.upvotes + post.views,
  };
};

export const getFormattedCampaign = ({
  spend,
  budget,
  startedAt,
  endedAt,
  ...campaign
}: GetCampaignResponse): GQLPromotedPost => ({
  ...campaign,
  spend: usdToCores(parseFloat(spend)),
  budget: usdToCores(parseFloat(budget)),
  startedAt: debeziumTimeToDate(startedAt),
  endedAt: debeziumTimeToDate(endedAt),
});

export interface BoostedPostStats
  extends Pick<GetCampaignListResponse, 'clicks' | 'impressions' | 'users'> {
  engagements: number;
  totalSpend: number;
}

export interface BoostedPostConnection extends Connection<GQLBoostedPost> {
  stats?: BoostedPostStats;
}

export const consolidateCampaignsWithPosts = async (
  campaigns: GetCampaignResponse[],
  con: ConnectionManager,
): Promise<GQLBoostedPost[]> => {
  const ids = campaigns.map(({ postId }) => postId);
  const builder = getBoostedPostBuilder(con);
  const postAlias = 'p1';
  const posts = await builder
    .where(`"${postAlias}".id IN (:...ids)`, { ids })
    .getRawMany<GetBoostedPost>();
  const mapped = posts.reduce(
    (map, post) => ({ ...map, [post.id]: post }),
    {} as Record<string, GetBoostedPost>,
  );

  return campaigns.map((campaign) => ({
    campaign: getFormattedCampaign(campaign),
    post: getFormattedBoostedPost(mapped[campaign.postId]),
  }));
};

interface TotalEngagements
  extends Pick<Post, 'views' | 'comments' | 'upvotes'> {}

export const getTotalEngagements = async (
  con: ConnectionManager,
  postIds: string[],
): Promise<number> => {
  const builder = con.getRepository(Post).createQueryBuilder();

  const engagements = await builder
    .select('SUM(upvotes)::int', 'upvotes')
    .addSelect('SUM(comments)::int', 'comments')
    .addSelect('SUM(views)::int', 'views')
    .where({ id: In(postIds) })
    .getRawOne<TotalEngagements>();

  return (
    (engagements?.upvotes || 0) +
    (engagements?.comments || 0) +
    (engagements?.views || 0)
  );
};

export const generateBoostEmailUpdate: TemplateDataFunc = async (
  con,
  user,
  notification,
) => {
  const campaign = await skadiApiClientV1.getCampaignById({
    campaignId: notification.referenceId!,
    userId: user.id,
  });

  if (!campaign) {
    return null;
  }

  const post = await con.getRepository(Post).findOne({
    where: { id: campaign.postId },
  });

  if (!post) {
    return null;
  }

  const sharedPost = await (post.type === PostType.Share
    ? con.getRepository(ArticlePost).findOne({
        where: { id: (post as SharePost).sharedPostId },
        select: ['title', 'image', 'slug'],
      })
    : Promise.resolve(null));

  const title = truncatePostToTweet(post || sharedPost);
  const engagement = post.views + post.upvotes + post.comments;

  return {
    start_date: formatMailDate(debeziumTimeToDate(campaign.startedAt)),
    end_date: formatMailDate(debeziumTimeToDate(campaign.endedAt)),
    impressions: largeNumberFormat(campaign.impressions),
    clicks: largeNumberFormat(campaign.clicks),
    engagement: largeNumberFormat(engagement),
    post_link: getDiscussionLink(post.slug),
    analytics_link: addNotificationEmailUtm(
      notification.targetUrl,
      notification.type,
    ),
    post_image: sharedPost?.image || (post as FreeformPost).image,
    post_title: title,
  };
};

export const getAdjustedReach = (value: number) => {
  // We do plus-minus 8% of the generated value
  const difference = Math.floor(value * 0.08);
  const min = Math.max(value - difference, 0);
  const estimatedReach = {
    min,
    max: Math.max(value + difference, min),
  };

  return estimatedReach;
};

export const startCampaignPost = async (props: StartCampaignMutationArgs) => {
  const { ctx, args } = props;
  const { value: postId } = args;
  const post = await validatePostBoostPermissions(ctx, postId);
  checkPostAlreadyBoosted(post);

  const request = await ctx.con.transaction(async (manager) => {
    const id = randomUUID();
    const creativeId = randomUUID();
    const { budget, duration } = args;
    const total = budget * duration;
    const userId = ctx.userId;
    const endedAt = addDays(new Date(), duration);

    const campaign = await manager.getRepository(CampaignPost).save(
      manager.getRepository(CampaignPost).create({
        id,
        creativeId,
        flags: {
          budget: total,
          spend: 0,
          users: 0,
          clicks: 0,
          impressions: 0,
        },
        userId,
        referenceId: postId,
        state: CampaignState.Active,
        endedAt,
        postId,
        type: CampaignType.Post,
      }),
    );

    const campaignId = campaign.id;
    const tags = await getPostTags(manager, postId);

    await skadiApiClientV2.startCampaign(campaign, tags);

    await manager
      .getRepository(Post)
      .update(
        { id: postId },
        { flags: updateFlagsStatement<Post>({ campaignId: id }) },
      );

    const userTransaction = await manager.getRepository(UserTransaction).save(
      manager.getRepository(UserTransaction).create({
        id: randomUUID(),
        processor: UserTransactionProcessor.Njord,
        receiverId: systemUser.id,
        status: UserTransactionStatus.Success,
        productId: null,
        senderId: userId,
        value: total,
        valueIncFees: 0,
        fee: 0,
        request: ctx.requestMeta,
        flags: { note: `Post Boost started` },
        referenceId: campaignId,
        referenceType: UserTransactionType.PostBoost,
      }),
    );

    return await startCampaignTransferCores({
      ctx,
      manager,
      campaignId,
      userTransaction,
    });
  });

  return request.transaction;
};

export const stopCampaignPost = async ({
  campaign,
  ctx,
}: StopCampaignProps) => {
  const { id: campaignId, userId, referenceId } = campaign;

  const { budget } = await skadiApiClientV2.cancelCampaign({
    campaignId,
    userId,
  });

  const result = await ctx.con.transaction(async (manager) => {
    const toRefund = parseFloat(budget);

    await manager
      .getRepository(Post)
      .update(
        { id: referenceId },
        { flags: updateFlagsStatement<Post>({ campaignId: null }) },
      );

    const userTransaction = await manager.getRepository(UserTransaction).save(
      manager.getRepository(UserTransaction).create({
        id: randomUUID(),
        processor: UserTransactionProcessor.Njord,
        receiverId: userId,
        status: UserTransactionStatus.Success,
        productId: null,
        senderId: systemUser.id,
        value: usdToCores(toRefund),
        valueIncFees: 0,
        fee: 0,
        flags: { note: `Post Boost refund` },
        referenceId: campaignId,
        referenceType: UserTransactionType.PostBoost,
      }),
    );

    return await stopCampaignTransferCores({
      ctx,
      manager,
      campaignId,
      userTransaction,
    });
  });

  return result.transaction;
};

export const getPostTags = async (con: ConnectionManager, postId: string) => {
  const builder = con.getRepository(Post).createQueryBuilder('p1');
  const subquery = builder
    .subQuery()
    .select(`COALESCE(p2."tagsStr", '')`)
    .from(Post, 'p2')
    .where('p2.id = p1."sharedPostId"')
    .getQuery();

  const result = await builder
    .select(`COALESCE(p1."tagsStr", '')`, 'tagsStr')
    .addSelect(`(${subquery})`, 'sharedTagsStr')
    .where('p1.id = :id', { id: postId })
    .getRawOne<{ tagsStr: string; sharedTagsStr: string }>();

  const tags1 = (result?.tagsStr ?? '').split(',');
  const tags2 = (result?.sharedTagsStr ?? '').split(',');
  const list = [...tags1, ...tags2].filter((tag) => tag.trim().length > 0);

  return Array.from(new Set(list));
};
