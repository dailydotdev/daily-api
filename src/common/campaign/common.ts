import { ValidationError } from 'apollo-server-errors';
import {
  Campaign,
  CampaignPost,
  CampaignType,
  type ConnectionManager,
} from '../../entity';
import { UserTransaction } from '../../entity/user/UserTransaction';
import { parseBigInt } from '../utils';
import { TransferError } from '../../errors';
import { transferCores, throwUserTransactionError, getBalance } from '../njord';
import type { AuthContext } from '../../Context';
import type { EntityManager } from 'typeorm';
import { CAMPAIGN_VALIDATION_SCHEMA } from '../schema/campaigns';
import { getSourceTags } from './source';
import { generatePostBoostEmail, getPostTags } from './post';
import type { NotificationBuilder } from '../../notifications/builder';
import { NotificationIcon } from '../../notifications/icons';
import { notificationsLink } from '../links';
import type { NotificationCampaignContext } from '../../notifications';
import type { TemplateDataFunc } from '../../workers/newNotificationV2Mail';
import { queryReadReplica } from '../queryReadReplica';

export interface StartCampaignArgs {
  value: string;
  duration: number;
  budget: number;
}

export interface StartCampaignMutationArgs {
  ctx: AuthContext;
  args: StartCampaignArgs;
}

export const validateCampaignArgs = (
  args: Pick<StartCampaignArgs, 'budget' | 'duration'>,
) => {
  const result = CAMPAIGN_VALIDATION_SCHEMA.safeParse(args);

  if (result.error) {
    throw new ValidationError(result.error.issues[0].message);
  }
};

interface StartCampaignTransferCoresProps {
  ctx: AuthContext;
  campaignId: string;
  userTransaction: UserTransaction;
  manager: EntityManager;
}

export const startCampaignTransferCores = async ({
  ctx,
  campaignId,
  userTransaction,
  manager,
}: StartCampaignTransferCoresProps) => {
  if (ctx.isTeamMember) {
    return {
      transaction: {
        referenceId: campaignId,
        transactionId: userTransaction.id,
        balance: { amount: (await getBalance({ userId: ctx.userId })).amount },
      },
    };
  }

  try {
    const transfer = await transferCores({
      ctx,
      transaction: userTransaction,
      entityManager: manager,
    });

    return {
      transfer,
      transaction: {
        referenceId: campaignId,
        transactionId: userTransaction.id,
        balance: {
          amount: parseBigInt(transfer.senderBalance?.newBalance),
        },
      },
    };
  } catch (error) {
    if (error instanceof TransferError) {
      await throwUserTransactionError({
        ctx,
        entityManager: manager,
        error,
        transaction: userTransaction,
      });
    }

    throw error;
  }
};

export const stopCampaignTransferCores = async ({
  ctx,
  campaignId,
  userTransaction,
  manager,
}: StartCampaignTransferCoresProps) => {
  if (ctx.isTeamMember) {
    return {
      transaction: {
        referenceId: campaignId,
        transactionId: userTransaction.id,
        balance: { amount: (await getBalance({ userId: ctx.userId })).amount },
      },
    };
  }

  try {
    const transfer = await transferCores({
      ctx,
      transaction: userTransaction,
      entityManager: manager,
    });

    return {
      transfer,
      transaction: {
        referenceId: campaignId,
        transactionId: userTransaction.id,
        balance: {
          amount: parseBigInt(transfer.receiverBalance?.newBalance),
        },
      },
    };
  } catch (error) {
    if (error instanceof TransferError) {
      await throwUserTransactionError({
        ctx,
        entityManager: manager,
        error,
        transaction: userTransaction,
      });
    }

    throw error;
  }
};

export interface StopCampaignProps {
  campaign: Campaign;
  ctx: AuthContext;
}

export const getReferenceTags = (
  con: ConnectionManager,
  type: CampaignType,
  referenceId: string,
) => {
  switch (type) {
    case CampaignType.Post:
      return getPostTags(con, referenceId);
    case CampaignType.Squad:
      return getSourceTags(con, referenceId);
    default:
      throw new ValidationError('Unknown campaign type to estimate reach');
  }
};

export enum CampaignUpdateEvent {
  Started = 'CAMPAIGN_STARTED',
  Completed = 'CAMPAIGN_COMPLETED',
  StatsUpdated = 'STATS_UPDATED',
  BudgetUpdated = 'BUDGET_UPDATED',
}

export interface CampaignCompleted {
  budget: string;
}

export interface CampaignStatsUpdate {
  impressions: number;
  clicks: number;
  unique_users: number;
}

export interface CampaignStateUpdate {
  budget: string; // used budget
}

export interface CampaignUpdateEventArgs {
  campaignId: string;
  event: CampaignUpdateEvent;
  unique_users: number;
  data: CampaignCompleted | CampaignStatsUpdate | CampaignStateUpdate;
  d_update: number;
}

export const generateCampaignCompletedNotification = (
  builder: NotificationBuilder,
  ctx: NotificationCampaignContext,
) => {
  const { campaign, source, event, user } = ctx;

  const nb = builder
    .icon(NotificationIcon.DailyDev)
    .referenceCampaign(ctx)
    .targetUrl(notificationsLink)
    .setTargetUrlParameter([['c_id', campaign.id]])
    .uniqueKey(`${campaign.id}-${user.id}-${event}`);

  switch (campaign.type) {
    case CampaignType.Post:
      return nb.avatarUser(user);
    case CampaignType.Squad:
      if (!source) {
        throw new Error(
          `Can't generate Squad Campaign Notification without the Squad`,
        );
      }
      return nb.avatarSource(source);
    default:
      throw new Error(
        `Unable to generate notification for unknown type: ${campaign.type}`,
      );
  }
};

export const generateCampaignCompletedEmail: TemplateDataFunc = async (
  con,
  user,
  notification,
) => {
  const campaign = await con
    .getRepository(Campaign)
    .findOneBy({ id: notification.referenceId });

  if (!campaign) {
    return null;
  }

  switch (campaign.type) {
    case CampaignType.Post:
      return generatePostBoostEmail({
        con,
        postId: campaign.referenceId,
        notification,
        campaign,
      });
    // TODO: MI-1007 - generate email once template id is provided
    default:
      return null;
  }
};

export interface UserCampaignStats {
  impressions: number;
  clicks: number;
  spend: number;
  users: number;
}

export const getUserCampaignStats = async (
  ctx: AuthContext,
): Promise<UserCampaignStats> => {
  const result = await queryReadReplica(ctx.con, ({ queryRunner }) =>
    queryRunner.manager
      .getRepository(CampaignPost)
      .createQueryBuilder('c')
      .select(`SUM(COALESCE((c.flags->>'impressions')::int, 0))`, 'impressions')
      .addSelect(`SUM(COALESCE((c.flags->>'users')::int, 0))`, 'users')
      .addSelect(`SUM(COALESCE((c.flags->>'clicks')::int, 0))`, 'clicks')
      .addSelect(`SUM(COALESCE((c.flags->>'spend')::int, 0))`, 'spend')
      .where(`c."userId" = :user`, { user: ctx.userId })
      .getRawOne(),
  );

  return {
    clicks: result.clicks ?? 0,
    impressions: result.impressions ?? 0,
    users: result.users ?? 0,
    spend: result.spend ?? 0,
  };
};
