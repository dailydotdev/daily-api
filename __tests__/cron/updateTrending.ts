import cron from '../../src/cron/updateTrending';
import { expectSuccessfulCron, saveFixtures } from '../helpers';
import { ArticlePost, Post, Source, View } from '../../src/entity';
import { sourcesFixture } from '../fixture/source';
import { postsFixture, sharedPostsFixture } from '../fixture/post';

import { DeepPartial } from 'typeorm/common/DeepPartial';
import { DataSource } from 'typeorm';
import createOrGetConnection from '../../src/db';

let con: DataSource;

beforeAll(async () => {
  con = await createOrGetConnection();
});

beforeEach(async () => {
  await saveFixtures(con, Source, sourcesFixture);
  await saveFixtures(con, ArticlePost, postsFixture);
  await saveFixtures(con, ArticlePost, sharedPostsFixture);
});

const addViewsToPost = async (
  postId: string,
  count: number,
  now: Date,
): Promise<void> => {
  await con.getRepository(View).save(
    [...new Array(count)].map(
      (val, index): DeepPartial<View> => ({
        postId,
        userId: index.toString(),
        timestamp: new Date(now.getTime() - Math.random() * 20 * 60 * 1000),
      }),
    ),
  );
};

it('should update the trending score of the relevant articles', async () => {
  const now = new Date();
  const halfHour = new Date(now.getTime() - 30 * 60 * 1000);
  const hour = new Date(now.getTime() - 60 * 60 * 1000);
  await Promise.all([
    con.getRepository(Post).update(postsFixture[0].id, { createdAt: hour }),
    con.getRepository(Post).update(postsFixture[2].id, { createdAt: hour }),
    addViewsToPost(postsFixture[0].id, 80, now),
    addViewsToPost(postsFixture[0].id, 20, halfHour),
    addViewsToPost(postsFixture[1].id, 10, now),
    addViewsToPost(postsFixture[1].id, 3, halfHour),
    addViewsToPost(postsFixture[2].id, 170, now),
    addViewsToPost(postsFixture[2].id, 30, halfHour),
  ]);
  await expectSuccessfulCron(cron);
  const posts = await con
    .getRepository(Post)
    .find({ select: ['id', 'trending'], order: { id: 'ASC' } });
  expect(posts).toMatchSnapshot();
  const trendingPost = await con.getRepository(Post).findOne({
    select: ['id', 'lastTrending'],
    where: { id: postsFixture[0].id },
  });
  expect(trendingPost.lastTrending).toBeTruthy();
});

it('should set trending to null when no longer trending', async () => {
  const now = new Date();
  await Promise.all([
    addViewsToPost(
      postsFixture[0].id,
      100,
      new Date(now.getTime() - 120 * 60 * 1000),
    ),
    con.getRepository(Post).update(postsFixture[0].id, {
      trending: 150,
      lastTrending: new Date(now.getTime() - 60 * 60 * 1000),
    }),
  ]);
  await expectSuccessfulCron(cron);
  const post = await con.getRepository(Post).findOne({
    select: ['id', 'trending'],
    where: { id: postsFixture[0].id },
  });
  expect(post.trending).toEqual(null);
});
