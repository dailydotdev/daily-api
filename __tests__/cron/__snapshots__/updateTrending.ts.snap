// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should update the trending score of the relevant articles 1`] = `
Array [
  Post {
    "id": "p1",
    "trending": 100,
  },
  Post {
    "id": "p2",
    "trending": null,
  },
  Post {
    "id": "p3",
    "trending": 200,
  },
  Post {
    "id": "p4",
    "trending": null,
  },
  Post {
    "id": "p5",
    "trending": null,
  },
  Post {
    "id": "p6",
    "trending": null,
  },
  Post {
    "id": "p7",
    "trending": null,
  },
  Post {
    "id": "squadP1",
    "trending": null,
  },
  Post {
    "id": "yt2",
    "trending": null,
  },
]
`;
